File: dtx/runtime/csharp/src/Dtx/Fragment.cs
namespace LLVM.DTX;

public static class Fragment
{
    public static byte[] BuildSelectorPayload(string selector, AuxList args)
    {
        var auxBytes = args.Encode();
        var selectorBytes = KeyedArchiver.ArchiveRoot(NSObject.String(selector));
        return BuildPayload(PayloadFlag.Selector, auxBytes, selectorBytes);
    }

    public static byte[] BuildObjectPayload(NSObject value, PayloadFlag flag = PayloadFlag.Reply) =>
        BuildPayload(flag, Array.Empty<byte>(), KeyedArchiver.ArchiveRoot(value));

    public static byte[] BuildFrame(MessageHeader header, byte[] payload)
    {
        using var stream = new MemoryStream();
        var headerBytes = header.Encode();
        stream.Write(headerBytes, 0, headerBytes.Length);
        stream.Write(payload, 0, payload.Length);
        return stream.ToArray();
    }

    public static NSObject ParseObjectPayload(byte[] payload)
    {
        var header = PayloadHeader.Decode(payload);
        if (header.Flags != (uint)PayloadFlag.Reply &&
            header.Flags != (uint)PayloadFlag.Object &&
            header.Flags != (uint)PayloadFlag.AsyncObject)
            throw new DtxException("Payload is not an object reply.");
        if (header.AuxiliaryLength != 0)
            throw new DtxException("Object payload unexpectedly contains aux bytes.");
        if (16 + checked((int)header.TotalLength) != payload.Length)
            throw new DtxException("Object payload length mismatch.");
        if (header.TotalLength == 0)
            return NSObject.Null;
        return KeyedArchiver.UnarchiveRoot(payload, 16, checked((int)header.TotalLength));
    }

    private static byte[] BuildPayload(PayloadFlag flag, byte[] auxBytes, byte[] bodyBytes)
    {
        var header = new PayloadHeader(
            (uint)flag,
            checked((uint)auxBytes.Length),
            checked((ulong)(auxBytes.Length + bodyBytes.Length)));
        using var stream = new MemoryStream();
        var headerBytes = header.Encode();
        stream.Write(headerBytes, 0, headerBytes.Length);
        stream.Write(auxBytes, 0, auxBytes.Length);
        stream.Write(bodyBytes, 0, bodyBytes.Length);
        return stream.ToArray();
    }
}

File: dtx/runtime/kotlin/src/main/kotlin/org/llvm/dtx/Fragment.kt
package org.llvm.dtx

import java.io.ByteArrayOutputStream

object Fragment {
    fun buildSelectorPayload(selector: String, args: AuxList): ByteArray {
        val auxBytes = args.encode()
        val selectorBytes = KeyedArchiver.archiveRoot(NSObject.string(selector))
        return buildPayload(PayloadFlag.Selector, auxBytes, selectorBytes)
    }

    fun buildObjectPayload(value: NSObject, flag: PayloadFlag = PayloadFlag.Reply): ByteArray =
        buildPayload(flag, ByteArray(0), KeyedArchiver.archiveRoot(value))

    fun buildFrame(header: MessageHeader, payload: ByteArray): ByteArray {
        val out = ByteArrayOutputStream()
        out.write(header.encode())
        out.write(payload)
        return out.toByteArray()
    }

    fun parseObjectPayload(payload: ByteArray): NSObject {
        val header = PayloadHeader.decode(payload)
        if (header.flags != PayloadFlag.Reply.value &&
            header.flags != PayloadFlag.Object.value &&
            header.flags != PayloadFlag.AsyncObject.value)
            throw DtxException("Payload is not an object reply.")
        if (header.auxiliaryLength != 0)
            throw DtxException("Object payload unexpectedly contains aux bytes.")
        if (16 + header.totalLength.toInt() != payload.size)
            throw DtxException("Object payload length mismatch.")
        if (header.totalLength == 0L)
            return NSObject.Null
        return KeyedArchiver.unarchiveRoot(payload, 16, header.totalLength.toInt())
    }

    private fun buildPayload(flag: PayloadFlag, auxBytes: ByteArray, bodyBytes: ByteArray): ByteArray {
        val header = PayloadHeader(flag.value, auxBytes.size, (auxBytes.size + bodyBytes.size).toLong())
        val out = ByteArrayOutputStream()
        out.write(header.encode())
        out.write(auxBytes)
        out.write(bodyBytes)
        return out.toByteArray()
    }
}
